src/Algebra/Polynomial/Morphism/Morphism.agda

```diff
@@ -0,0 +1,510 @@
+-- Polynomial Morphisms
+open import Algebra.Bundles
+  using (CommutativeRing ; CommutativeSemiring; Ring;
+         SemiringWithoutAnnihilatingZero; Semiring)
+
+module ALgebra.Polynomial.Morphism.Morphism
+  {ℓ₁ ℓ₂}
+  (R : CommutativeRing ℓ₁ ℓ₂)
+  where
+
+open import Algebra.Bundles.Raw using (RawRing)
+open import Algebra.Morphism.Bundles
+  using (RingHomomorphism ; MagmaHomomorphism ; MonoidHomomorphism)
+open import Algebra.Morphism.Structures
+  using (IsRingHomomorphism ; IsMagmaHomomorphism ; IsSemiringHomomorphism;
+         IsNearSemiringHomomorphism ; IsMonoidHomomorphism)
+open import Algebra.Polynomial.Poly.Base2 R
+  using (Poly; []≈ ; ≈[]; cons ; cast-lem ; consˡ ; consʳ)
+  renaming ( sym to P-sym ; _·_ to _·P_
+           ; _+_ to _+P_ ; _*_ to _*P_; _≈_ to _≈P_; -_ to -P_)
+open import Algebra.Polynomial.Poly.Properties2 R
+  using (cast-p≈p; IsZero⇒≈0; ≈0⇒IsZero; a·p≈a∷[]*p ; *-distrib-shift)
+  renaming  (+-identityˡ to +-P-identityˡ ; +-identityʳ to +-P-identityʳ
+            ; zeroˡ to P-zeroˡ; zeroʳ to P-zeroʳ ; *-comm to *-P-comm )
+open import Algebra.Polynomial.Base2 R
+open import Algebra.Polynomial.Properties2 R as P
+  using (+-identityˡ; +-identityʳ; +-*-commutativeRing; *-zeroˡ ; *-zeroʳ)
+
+import Algebra.Definitions.RawSemiring as Def
+import Algebra.Definitions as D 
+open import Algebra.Properties.Group
+open import Algebra.Properties.Ring
+open import Data.Nat.Base as ℕ using (ℕ; suc; pred; _⊔_)
+open import Data.Vec.Base as Vec using (Vec; []; _∷_)
+open import Data.Vec.Relation.Unary.All
+open import Data.Product.Base using (proj₁ ; proj₂)
+import Level as L
+import Relation.Binary.Morphism.Structures as S
+  using (IsRelHomomorphism)
+open import Relation.Binary.PropositionalEquality.Core using (cong)
+
+module R = CommutativeRing R
```

Comment:
I think that `module A = CommutativeRing R` and then using `A.foo` below would be much nicer than renaming.